### PR TITLE
feat(concurrent-backup-restore): implement

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/4558

### CODE REVIEW

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2638/testReport/tests/test_settings/

**Comment by c3y1huang** [comment]
Test result https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2647/testReport/tests/test_settings/

File: controller/engine_controller.go [hunk 1]
```diff
@@ -1002,6 +1030,29 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
...
+		m.restoringCounter.IncreaseCount()
+		m.restoringCounterAquired = aquire
+		return
+	}
+
+	if !aquire && m.restoringCounterAquired {
+		m.restoringCounter.DecreaseCount()
+		m.restoringCounterAquired = aquire
+		return
+	}
```
Comment (derekbit) [inline controller/engine_controller.go:1053]:
```suggestion
	if !aquire && m.restoringCounterAquired {
		m.restoringCounter.DecreaseCount()
		m.restoringCounterAquired = aquire
	}
```

File: controller/engine_controller.go [hunk 2]
```diff
@@ -973,11 +989,23 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
 			m.logger.Debug("Cannot restore the backup for engine since it is still in the backoff window")
 			return nil
 		}
+
+		if isReachedLimit, err := m.isReachedConcurrentVolumeBackupRestoreLimit(); err != nil {
+			m.logger.WithError(err).Warn("Failed to check concurrent volume backup restore limit")
+			return nil
+
+		} else if isReachedLimit {
+			m.logger.Debugf("Reached the concurrent volume backup restore limit of %v, retry later", types.SettingNameConcurrentBackupRestorePerNodeLimit)
```
Comment (derekbit) [inline controller/engine_controller.go:998]: Should it be warning level message?

**Review by innobead — CHANGES_REQUESTED** [review]

File: controller/engine_controller.go (hunk 2 above)
Comment (derekbit) [reply controller/engine_controller.go:998]: I see. Just thought if we can easily know why the backup does not take place from log.

File: controller/engine_controller.go [hunk 3]
```diff
@@ -973,11 +977,24 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
...
+			m.restoreBackoff.Next(engine.Name, time.Now())
+			return nil
+		}
```
Comment (innobead) [inline controller/engine_controller.go:989]: You can add this part to `aquireRestoringCounter`, but bypass this check if the volume is DR.

File: controller/engine_controller.go [hunk 4]
```diff
@@ -1002,6 +1009,24 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
...
+
+	volume, err := m.ds.GetVolumeRO(engine.Spec.VolumeName)
+	if err != nil {
+		m.logger.Warnf("Failed to get volume %v for restoring counter", engine.Spec.VolumeName)
+		return
+	}
+	isDRVolume := volume.Status.IsStandby
+	if isDRVolume {
+		return
+	}
```
Comment (innobead) [inline controller/engine_controller.go:1036]: Move this to the top, because need to bypass all counting if the volume is DR.

**Review by innobead — CHANGES_REQUESTED** [review]

**Review by innobead — DISMISSED** [review]
LGTM

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2664/

File: controller/engine_controller.go (hunk 3 above)
Comment (c3y1huang) [reply controller/engine_controller.go:989]: We need to return this, so keep this as it is.

File: controller/engine_controller.go (hunk 4 above)
Comment (c3y1huang) [reply controller/engine_controller.go]: Moved this to the caller so it will only be checked once.

File: controller/engine_controller.go (hunk 3 above)
Comment (innobead) [reply controller/engine_controller.go:989]: You are right!

File: controller/engine_controller.go [hunk 5]
```diff
@@ -973,11 +977,31 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
...
+
+		if isReachedLimit, err := m.isReachedConcurrentVolumeBackupRestoreLimit(); err != nil {
+			m.logger.WithError(err).Warn("Failed to check concurrent volume backup restore limit")
+			return nil
+
+		} else if isReachedLimit {
+			m.logger.Debugf("Reached the concurrent volume backup restore limit of %v, retry later", types.SettingNameConcurrentBackupRestorePerNodeLimit)
+			m.restoreBackoff.Next(engine.Name, time.Now())
+			return nil
+		}
```
Comment (innobead) [inline controller/engine_controller.go:989]: This part should be checked only if the volume is not DR, right?

Should move into the below `if !isDRVolume {`

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2665/console

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2665/

**Review by innobead — APPROVED** [review]

**Comment by shuo-wu** [comment]
Sorry for the late comment, I think there may be one issue:
The restore actually happens inside a replica process. And IIRC, Longhorn manager does not guarantee the engine and the restoring replica are on the same node. In other words, this PR cannot limit the restoring replica count on a node.

**Comment by innobead** [comment]
> Sorry for the late comment, I think there may be one issue: The restore actually happens inside a replica process. And IIRC, Longhorn manager does not guarantee the engine and the restoring replica are on the same node. In other words, this PR cannot limit the restoring replica count on a node.

You are right. Here, @c3y1huang 's implementation just limits the count of engines but the real processing unit is each replica which is likely not on the same node. However, this is still can take some effect from the cluster perspective because this concurrent limitation will be accumulated. 

This can be improved in the future to make the limitation precise for each node.

**Pull request merged**